### PR TITLE
Expand-the-VMware-Integration

### PR DESCRIPTION
**Pull request by mihaitusa**
Customizing Windows Guest, Customize Linux Guest, Mount Tools && Enhancements on: Get VM Details, CreateVM, CloneVM.

Signed-off-by: Mihail-Cristian Tusa <mihail-cristian.tusa@hpe.com>

### CODE REVIEW

File: score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java [hunk 1]
```diff
@@ -35,6 +35,10 @@
     public static final String ATAPI_CONTROLLER_CAPACITY_MAXED_OUT = "The IDE controller on the vm has maxed out its " +
             "capacity. Please add an additional IDE controller";
 
+    public static final String NOT_BYTE = "The input value must be a byte number.";
     public static final String NOT_INTEGER = "The input value must be a int number.";
     public static final String NOT_LONG = "The input value must be a long number.";
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java]: "The input value must be a long number"... does the user know the range of long? Do negative numbers count as valid values ? 

I suggest changing the messages and explicitly say what the maximum value is for each.

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+            results = new ResponseHelper().getResultsMap(connectionResources, task,
+                    "Success: The [" + vmInputs.getVirtualMachineName() + "] VM was successfully customized. The taskId is: " +
+                            task.getValue(), "Failure: The [" + vmInputs.getVirtualMachineName() + "] VM could not be customized.");
+        } else {
+            results = new HashMap<>();
+            ResponseUtils.setResults(results, "Could not find the [" + vmInputs.getVirtualMachineName() + "] VM.",
+                    Outputs.RETURN_CODE_FAILURE);
+        }
+        connectionResources.getConnection().disconnect();
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java]: if exception is thrown before line 56, connection will not be disconnected. Correct way is to use try with resources.

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java]: same thing here: use try with resources

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     * by the inputs provided.
+     *
+     * @param httpInputs  Object that has all the inputs necessary to made a connection to data center
+     * @param vmInputs    Object that has all the specific inputs necessary to identify the targeted virtual machine
+     * @param guestInputs Object that has all specific inputs necessary to customize specified virtual machine
+     * @return Map with String as key and value that contains returnCode of the operation, success message with task id
+     * of the execution or failure message and the exception if there is one
+     * @throws Exception
+     */
+    public Map<String, String> customizeWinVM(HttpInputs httpInputs, VmInputs vmInputs, GuestInputs guestInputs)
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/GuestService.java]: refactor this and customizeLinVM in at least 3 separate methods:

1) getVmMor()
2) getCustomizationSpec
3) customizeVMTask

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java [hunk 5]
```diff
@@ -45,13 +45,13 @@
...
+        ManagedObjectReference task = connectionResources.getVimPortType()
                 .createVMTask(vmFolderMor, vmConfigSpec, resourcePoolMor, hostMor);
 
-        Map<String, String> results = new HashMap<>();
-        setTaskResults(results, connectionResources, taskMor, "Success: Created [" + vmInputs.getVirtualMachineName() +
-                "] VM. The taskId is: " + taskMor.getValue(), "Failure: Creating [" + vmInputs.getVirtualMachineName() +
-                "] VM");
+        Map<String, String> results = new ResponseHelper().getResultsMap(connectionResources, task,
+                "Success: Created [" + vmInputs.getVirtualMachineName() + "] VM. The taskId is: " + task.getValue(),
+                "Failure: Creating [" + vmInputs.getVirtualMachineName() + "] VM");
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java]: Change failure error message to : Could not create ... VM

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/MorObjectHandler.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public class MorObjectHandler {
+    public ManagedObjectReference getVmMor(ConnectionResources connectionResources, String filter, String parameter) throws Exception {
+        ManagedObjectReference reference = connectionResources.getMorRootFolder();
+        return getSpecificMor(connectionResources, reference, filter, parameter);
+    }
+
+    public Map<String, ManagedObjectReference> getSpecificObjectsMap(ConnectionResources connectionResources, String objectType)
+            throws InvalidPropertyFaultMsg, RuntimeFaultFaultMsg {
+        return connectionResources.getGetMOREF().inContainerByType(connectionResources.getMorRootFolder(), objectType);
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/MorObjectHandler.java]: rename method name

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/MorObjectHandler.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public ManagedObjectReference getEnvironmentBrowser(ConnectionResources connectionResources, String filter)
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/MorObjectHandler.java]: add 'parameter' to method signature, and change 'getProperty' method call to (connectionResources, reference, filter, parameter)

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/ResponseHelper.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Mihai Tusa.
+ * 3/22/2016.
+ */
+public class ResponseHelper {
+    public Map<String, String> getResultsMap(ConnectionResources connectionResources, ManagedObjectReference task,
+                                             String successMessage, String failureMessage)
```
Comment (remusgolgot) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/helpers/ResponseHelper.java]: pass ConnectionResources and ManagedObjectReference objects in a constructor for ResponseHelper

**Comment by mihaitusa** [comment]
The review on this was already made by our colleague Remus which is the OO Content owner on VMware integration and it was an extensive one.

**Comment by remusgolgot** [comment]
everything looks good

**Comment by mihaitusa** [comment]
@remusgolgot  Thank you very much for your time and (sharp) observations!

**Comment by mihaitusa** [comment]
We will wait on merging/release to Maven Central until our QA team takes an extensive validation scenarios.

**Comment by mihaitusa** [comment]
https://github.com/CloudSlang/cloud-slang-content/issues/590 and https://github.com/CloudSlang/cloud-slang-content/issues/594

**Comment by mihaitusa** [comment]
@Bonczidai  Now (after an extensive QA validation process) I think we should release it to MC. WDYT?

**Comment by Bonczidai** [comment]
:+1: ok from my side

just be sure to validate by propagating locally to the CLI

**Comment by mbettan** [comment]
Let's merge guys, really need this content

**Comment by Bonczidai** [comment]
:ok_hand:

**Pull request merged**